daemon/control.c

```diff
@@ -69,96 +71,71 @@ struct control {
 		out.device_uuid = mem_strdup(cmld_get_device_uuid());
 	}
 
-	DEBUG("Opening and sending %s", log_file_name);
-	if (read_low_level) {
-		fp_low = open(log_file_name, O_RDONLY | O_NONBLOCK);
-		if (fp_low == -1) {
-			ERROR("Could not open %s", log_file_name);
-		} else {
-			file_is_open = true;
-		}
-	} else {
-		fp = fopen(log_file_name, "r");
-		line = NULL;
-		if (fp == NULL) {
-			ERROR("Could not open %s", log_file_name);
-		} else {
-			file_is_open = true;
-		}
-	}
+	DEBUG("Opening and sending %s", log_file_path);
 
-	while (file_is_open) {
-		if (read_low_level) {
-			/* The driver let's us read entry by entry */
-			bytes_read = read(fp_low, line_low, LOGGER_ENTRY_MAX_LEN);
-			if (bytes_read <= 0)
-				break;
-			char *first_string = line_low + HEADER_LENGTH;
-			size_t string_len = strlen(first_string);
-			char *second_string = line_low + HEADER_LENGTH + string_len + 1;
-			message.msg = mem_printf("%s/%s", first_string, second_string);
-		} else {
-			size_t len = 0;
-			bytes_read = getline(&line, &len, fp);
-			if (bytes_read == -1)
-				break;
-			message.msg = mem_strdup(line);
-		}
-		out.log_message = &message;
-		if (protobuf_send_message(fd, (ProtobufCMessage *)&out) < 0) {
-			ERROR_ERRNO("Could not finish sending %s", log_file_name);
-			skipped_lines = true;
-			break;
-		}
-		mem_free0(message.msg);
+	off_t filesize = file_size(log_file_path);
+
+	char *file_buf = file_read_new(log_file_path, (size_t)filesize);
+
+	if (file_buf) {
+		message.msg = mem_strdup(file_buf);
+		free(file_buf);
+	} else {
+		DEBUG("File %s could not be read to buffer.", log_file_path);
 	}
-	if (send_last_line_info) {
-		message.msg = mem_printf("Last line of log");
-		out.log_message = &message;
-		if (protobuf_send_message(fd, (ProtobufCMessage *)&out) < 0) {
-			ERROR("Could not sent last line info for %s", log_file_name);
-		}
-		mem_free0(message.msg);
+	out.log_message = &message;
+	if (protobuf_send_message(fd, (ProtobufCMessage *)&out) < 0) {
+		ERROR_ERRNO("Could not finish sending %s", log_file_path);
 	}
+	mem_free0(message.msg);
+
 	if (out.device_uuid != NULL) {
 		mem_free0(out.device_uuid);
 	}
-	if (file_is_open) {
-		if (read_low_level) {
-			close(fp_low);
-		} else {
-			fclose(fp);
-			if (line)
-				free(line);
-		}
-	}
 
-	if (!skipped_lines) {
-		DEBUG("Finished sending complete %s", log_file_name);
-	} else {
-		DEBUG("Finished sending %s; skipped lines", log_file_name);
-	}
+	DEBUG("Finished sending %s", log_file_path);
+}
+
+/**
+ * @brief callback for the dir_each function sending a file as LogMessage to the Controller
+ * @path: Expects path string without "/" at the end
+ * @return int 
+ */
+static int
+control_send_file_as_log_message_cb(const char *path, const char *file, void UNUSED *data)
+{
+	int *fd = (int *)data;
+	str_t *path_str = str_new(path);
+	str_append(path_str, "/");
+	str_append(path_str, file);
+	//get rid of const for later use in control_send_log
```

Comment:
instead of using str.c here, plz use mem_strdup when sending the protobuf message

As this callback is very small, i'd suggest moving the code of control_send_log_file here

Reply:
done, control_send_log_file is removed and included in the callback function